fix(ivy): Content Queries inheritance fix

AndrewKushnir opened the pull request:
Prior to this change `contentQueriesRefresh` functions that represent refresh logic for @ContentQuery list were not composable, which caused problems in case one Directive inherits another one and both of them contain Content Queries. Due to the fact that we used indices to reference queries in refresh function, results were placed into wrong Queries. In order to avoid that we no longer use indices to reference queries and instead maintain current content query index while iterating through them. This allows us to compose contentQueriesRefresh functions and make inheritance feature work with Content Queries.

This PR resolves issue FW-754.


## PR Type
What kind of change does this PR introduce?

- [x] Bugfix


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
fix(ivy): Content Queries inheritance fix

Prior to this change contentQueriesRefresh functions that represent refresh logic for @ContentQuery list were not composable, which caused problems in case one Directive inherits another one and both of them contain Content Queries. Due to the fact that we used indices to reference queries in refresh function, results were placed into wrong Queries. In order to avoid that we no longer use indices to reference queries and instead maintain current content query index while iterating through them. This allows us to compose contentQueriesRefresh functions and make inheritance feature work with Content Queries.

AndrewKushnir commented:
@kara I've rebased this PR on top of master branch (since View Queries fix was merged), so this PR now contains only relevant changes. Thank you.

AndrewKushnir committed:
fixup! fix(ivy): Content Queries inheritance fix

AndrewKushnir committed:
fixup! fix(ivy): Content Queries inheritance fix

JoostK commented:
OOC, are there plans to align `contentQueries`/`contentQueriesRefresh` with the `RenderFlags` approach as taken for the `viewQuery` function? View queries used to be executed from within the template function but moved into their own function in #24587, inheriting the `RenderFlags` pattern from the template function. Content queries however do not, which seems inconsistent.

AndrewKushnir committed:
fixup! fix(ivy): Content Queries inheritance fix

AndrewKushnir commented on packages/core/src/render3/query.ts:
```diff
@@ -419,4 +419,40 @@ export function loadViewQuery<T>(): T {
+ * @param read What to save in the query
+ * @returns QueryList<T>
+ */
+export function contentQuery<T>(
```
Wlcm :)

AndrewKushnir commented on packages/core/src/render3/state.ts:
```diff
@@ -271,6 +271,21 @@ export function setCurrentViewQueryIndex(value: number): void {
+ * Current index of a Content Query which needs to be processed next.
+ * We iterate over the list of Content Queries stored in LView and increment current query index.
+ */
+let contentQueryIndex: number = 0;
```
Good catch, thanks!

AndrewKushnir committed:
fixup! fix(ivy): Content Queries inheritance fix

AndrewKushnir committed:
fixup! fix(ivy): Content Queries inheritance fix

AndrewKushnir commented on packages/core/src/render3/query.ts:
```diff
@@ -419,4 +419,40 @@ export function loadViewQuery<T>(): T {
+    const lastSavedDirectiveIndex =
+        tView.contentQueries.length ? tView.contentQueries[tView.contentQueries.length - 2] : -1;
+    if (directiveIndex !== lastSavedDirectiveIndex) {
+      tViewContentQueries.push(directiveIndex, savedContentQueriesLength - 1);
```
Thanks @kara. I've updated the code to avoid storing query index as you proposed.

AndrewKushnir commented:
@kara thanks for the review. I've updated the code to addressed your comments. Could you please take another look?
I created a separate fixup commit (https://github.com/angular/angular/pull/28324/commits/0d2685a8e7e3ffba5b8762a19934fdfd624742e5) to implement your proposal on avoiding query index in `tView.contentQueries` - please let me know if that covers your proposal.

AndrewKushnir commented:
[Presubmit](http://test/OCL:230926480:BASE:230926548:1548441161347:9f4a7f25)

kara reviewed (changes requested):
Looks good, but one more quick thing - can you update docs for `TView.contentQueries` to remove mention of storing the query index? (it's in `interfaces/view.ts`)

AndrewKushnir committed:
fixup! fix(ivy): Content Queries inheritance fix

kara reviewed (approved):
LGTM

AndrewKushnir commented:
Caretaker: merge assistance is required: all tests pass (CircleCI and g3), but NG Bot got stuck and doesn't refresh "merge status" and "size" tasks.
